Create documentation workflow page

sarakonradi opened the pull request:
This PR creates a documentation workflow page. 

The page introduces the tools we use to generate and publish documentation (Sphinx, GitHub, Read the Docs).

It explains how to:
- Change the documentation 
- Set up a working environment
- Generate HTML with Sphinx
- Preview the content on Read the Docs

Currently, our User Documentation and Developer Space are separate from one another. @steffengraber and I discussed that, once the latter is available on Read the Docs, this workflow would be moved to the developer section.

steffengraber commented on doc/contents.rst:
```diff
@@ -21,6 +21,7 @@ Table of Contents
    getting_help
    ref_material/nest2_vs_3
    ref_material/pynest_apis
+   documentation_workflow
```
Wouldn't it be better to put it under 'GETTING INVOLED' after 'Contributing to NEST'?

sarakonradi replied:
Yes, that would make more sense!

steffengraber commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+
+.. code-block:: bash
+
+   cd git/nest-simulator/docs/_build/html
```
Maybe better '</path/to/nest_source>' than 'git/nest-simulator

sarakonradi replied:
Done.

steffengraber commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+
+.. code-block:: bash
+
+    cd git/nest-simulator/doc
```
Maybe better '</path/to/nest_source>' than 'git/nest-simulator

sarakonradi replied:
Done.

steffengraber commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+.. code-block:: bash
+
+   cd git/nest-simulator/docs/_build/html
+   Firefox FILENAME.html
```
use 'firefox'

sarakonradi replied:
Done.

steffengraber commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+
+4. If you add further changes to your files, repeat steps 1-3.
+
+Previewing on Read the Docs
```
This should only be optional

sarakonradi replied:
Done.

steffengraber commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+Creating pull request
+-------------------
+
+When you feel your documentation work is finished, you can create a pull request to the ``master`` branch of the NEST Source Code Repository. Your pull request will be reviewed by our NEST Documentation Team.
```
Maybe a link to https://nest-simulator.readthedocs.io/en/stable/contribute/index.html#make-changes-to-code-or-documentation would be nice.

sarakonradi replied:
You mean adding a sentence like ``For further details on how to contribute to NEST (new models, bug or documentation fixes), check out [LINK]``?

steffengraber commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+ ``~/.nestrc`` for an example setting firefox as browser.
+ Please note that the command helpdesk does not work if you have compiled
+ NEST with MPI support, but you have to enter the address of the helpdesk
+ (file:///install/path/share/doc/nest/index.html) manually into the browser.
```
`file:///</path/to/nest_install_dir>/share/doc/nest/index.html`

sarakonradi replied:
Done.

clinssen commented:
Could we standardise the reStructuredText heading levels? Case in point: the characters used in this PR's ``documentation_workflow.rst`` are incompatible with the ones in the new model documentation (#1374).

sarakonradi commented:
@steffengraber Thanks for your comments! I applied your suggestions. 

@clinssen Good that you are pointing that out. We can definitely standardize the heading levels throughout the NEST simulator documentation in a different PR. For now, I made the ``documentation_workflow.rst`` characters consistent with the model docs.

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+What you need to know
++++++++++++++++++++++
+
+We use **Sphinx** to generate documentation and **Read the Docs** to publish it.
```
A matter of philosophy: should Sphinx and Read the Docs be links to the respective websites (top level)?

sarakonradi replied:
Yes, that is a good idea! I have now linked Sphinx and Read the Docs. We should make sure such links remain up-to-date.

heplesser replied:
I assume that the top-level URLs will remain stable over a very long time.

sarakonradi replied:
That is true.

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+
+The documentation for NEST is contained in the ``docs`` directory within the `NEST source code repository <https://github.com/nest/nest-simulator>`_ on GitHub.
+
+Using **GitHub** allows the NEST simulator documentation to live alongside its code. Our developers benefit from smooth version control, with each release of NEST having its own documentation.
```
Isn't it Git that allows us to keep the versions under control, while GitHub is just the service we are currently using to run a Git server for us?

steffengraber replied:
Yes

sarakonradi replied:
I agree and have clarified that now.

sarakonradi replied:
I agree, we should clarify that. How about:

``We use Sphinx to generate documentation and Read the Docs to publish it.``

``The NEST simulator documentation lives alongside its code. It is contained in the docs directory within the NEST source code repository on GitHub.``

``We work with GitHub as the web-based hosting service for Git. Git allows us to keep our versions under control, with each release of NEST having its own documentation.``

steffengraber replied:
Ok for me.

sarakonradi replied:
Done.

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+
+Using **GitHub** allows the NEST simulator documentation to live alongside its code. Our developers benefit from smooth version control, with each release of NEST having its own documentation.
+
+.. image:: _static/img/documentation_workflow.png
```
I find the diagonal "Sphinx" banner in the figure confusing. There are thin arrows presumably indicating flow of information, but then the banned just cuts across. If Sphinx is used both on Github and RtD, we should have two Sphinx boxes with a size similar to other elements in the figure.

sarakonradi replied:
Yes, that would be clearer. I obtained the already existing image from Sciebo. @steffengraber Do you know who created this PNG file, or where the draft is located?

steffengraber replied:
@jessica-mitchell

heplesser replied:
@sarakonradi @jessica-mitchell If we cannot clarify/fix this within a few days, I would suggest to proceed with the current figure and create a follow-up issue about improving the figure.

sarakonradi replied:
@heplesser I received the original ideas and materials from @jessica-mitchell and created a new image. Sebastian L is also happy helping us from next week onward, so we can create a follow-up issue about improving the figure later on.

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+Changing the documentation
+++++++++++++++++++++++++++
+
+If you notice something out of place, you can submit an `issue <https://github.com/nest/nest-simulator/issues>`_ in our GitHub repository.
```
- "out of place" strictly speaking only means docs placed incorrectly. How about "any errors or other weaknesses in the documentation"
- instead of "you can submit" better "please submit"

sarakonradi replied:
Done.

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+
+.. code-block:: bash
+
+   conda env create -f conda_environment.yml
```
Should we tell the reader somewhere (e.g. on line 27 above, around "created a conda env") about this `conda_environment.yml` file? And maybe give it a more informative name, such as `nest_doc_conda_env.yml`?

steffengraber replied:
In #1411 I renamed the `.yml` to `nest_doc_conda_env.yml`. So, it should be: 

`    conda env create -f nest_doc_conda_env.yml`

sarakonradi replied:
Done.

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+   conda deactivate
+   conda remove --name doc --all
+
+Generating with Sphinx
```
Generating documentation with Sphinx

sarakonradi replied:
Done.

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+.. code-block:: bash
+
+   cd </path/to/nest_source>/doc/_build/html
+   firefox FILENAME.html
```
Must it be firefox?

steffengraber replied:
No. It can be any browser.

sarakonradi replied:
Changed ``firefox FILENAME.html`` to ``browser filename.html``

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+Previewing on Read the Docs (optional)
+++++++++++++++++++++++++++++++++++++++
+
+You can then preview your documentation on your Read the Docs account.
```
"You can then" implies the reader reads this page linearly.  How about "Proceed as follows to preview your version of the documentation on Read the Docs"?

sarakonradi replied:
Yes, that reads better. Done.

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+
+3. Save your changes.
+
+Rendering HTML
```
Should we move the generating page up, before the modification info?

sarakonradi replied:
I think the ``Generating documentation with Sphinx`` section fits well where it is. How about we move the sentence on Sphinx using reStructured text up? That could be added to ``What you need to know``.

heplesser replied:
@sarakonradi OK.

sarakonradi replied:
Done.

terhorstd replied:
The "Generating documentation with Sphinx" section is in a good place. Two thoughts on this though:

The sentence about reStructured text is fine up there, but I would repeat it with the the link also at the beginning of the "editing and creating pages" section. That's where you need the link and one may have skipped the colorful intro... or be too lazy to scroll all the way back up again.

I would also suggest to exchange the sections "Editing and creating pages" and "Rendering HTML". Reading a section on how to generate documentation I stumbled over "editing files" thinking "why do I need to edit files to build the documentation?. The section "Rendering HTML" is how you build the documentation yourself. Actually, going in and changing something is (read "should be") always the second step. Otherwise it is not possible to understand if a broken rendering is due to a problem introduced by the editing or due to other issues in the environment or a prior issues in the code base. When reading only the Rending HTML section, it sounds strange to have point 4. saying "add further changes", when points 1.-3. didn't do any.

My suggestion would be to exchange the sections, add a new point to editing and creating pages: "5. re-render the documentation as described above" and remove the point 4 from the rendering HTML section.

sarakonradi replied:
You are right, I see your point. Done.

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+
+.. code-block:: bash
+
+    cd git/nest-simulator/doc
```
As below (l 79), this should also be a placeholder path.

sarakonradi replied:
Done.

heplesser commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+
+To keep things simple, we have created a conda environment for you. Installing it will enable you to smoothly generate documentation for NEST.
+
+If you are using Linux and want to install a full development environment:
```
Works beautifully also on macOS, provided the yml file from #1411 is in place.

sarakonradi replied:
Yes.

heplesser commented:
Marked as Blocked, since #1411 must be merged first for the conda environment construction to work properly.

steffengraber commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+
+   conda env create -f conda_environment.yml
+   conda update -n base -c defaults conda
+   conda activate doc
```
In #1411 I renamed the environment to `nest-doc`. Now it should be:

   ` conda activate nest-doc`

sarakonradi replied:
Done.

steffengraber commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,152 @@
+.. code-block:: bash
+
+   conda deactivate
+   conda remove --name doc --all
```
See above:

    `conda remove --name nest-doc --all`

sarakonradi replied:
Done.

sarakonradi commented:
@heplesser @steffengraber Thanks for your comments! I made the changes and replied to your comments inline. 

Also added a point on updating cross-references to the ``Editing and creating pages`` section.

terhorstd commented:
unblocked since #1411 is merged.

jasperalbers commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,154 @@
+
+1. You can edit and/or add ``.rst`` files in the ``doc`` directory using your editor of choice.
+
+2. If you create a new page, open ``index.rst`` in the ``doc`` directory and add the file name under ``.. toctree::``. This will ensure it appears on the NEST simulator documentation's table of contents.
```
I have problems executing this step. In doc/index.rst, there is no .. toctree:: to which I can add the file name of my newly created .rst file. When I add .. toctree::my_new_page.rst myself, it does not add my new page to the table of contents. What am I missing here?

sarakonradi replied:
@jasperalbers Good that you noticed that. The ..toctree:: is in ``contents.rst``. Just corrected it.

jasperalbers commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,154 @@
+For **developer documentation**, we use `doxygen <http://doxygen.org/>`__
+comments extensively throughout NEST.
+After installing NEST, you can extract comments from the source code with
+``make doc`` and a doxygen folder with html files will be generated in the doc
```
Information where to execute `make doc` would be nice (I assume inside the folder doc).

jasperalbers replied:
`make doc` produces the following error for me:

➜ make doc
Running Sphinx v2.4.4
/bin/sh: doxygen: command not found
Wrote model documentation for model models/multimeter
Wrote model documentation for model models/spike_detector
Wrote model documentation for model models/weight_recorder
Wrote model documentation for model nestkernel/recording_backend_ascii
Wrote model documentation for model nestkernel/recording_backend_memory
Wrote model documentation for model nestkernel/recording_backend_screen
Wrote model documentation for model nestkernel/recording_backend_sionlib
/Users/work/opt/miniconda3/envs/nest-doc/lib/python3.8/site-packages/sphinx/util/docutils.py:285: RemovedInSphinx30Warning: function based directive support is now deprecated. Use class based directive instead.
  warnings.warn('function based directive support is now deprecated. '

Sphinx error:
Builder name doc not registered or available through entry point
make: *** [doc] Error 2

jasperalbers commented on doc/documentation_workflow.rst:
```diff
@@ -0,0 +1,154 @@
+
+ The command ``helpdesk()`` needs to know which browser to launch in order to display
+ the help pages. The browser is set as an option of helpdesk. Please see the file
+ ``~/.nestrc`` for an example setting firefox as browser.
```
Could be extended to also support Safari (default browser on MacOS) if that is desired.

sarakonradi commented:
After speaking to @steffengraber, we agreed that this page should focus on user documentation rather than developer documentation. 

Therefore, I:
- Divided the workflow files into ``user_documentation_workflow.rst`` and ``developer_documentation_workflow.rst``
- Moved the developer documentation workflow description into the latter
- Clarified the instructions on ``make doc``, as pointed out by @jasperalbers 
- Linked the developer workflow in the ``user_documentation_workflow``'s ``Note``, and vice versa

@jasperalbers You require Doxygen in order to execute ``make doc``. I have now added information on how to install Doxygen and where to enter ``make doc`` (see ``developer_documentation_workflow.rst``). Let me know if it works!

jasperalbers commented on doc/documentation_workflow/developer_documentation_workflow.rst:
```diff
@@ -0,0 +1,70 @@
+
+.. code-block::
+
+   sudo apt install doxygen
```
MacOS: `brew install doxygen` (assuming that brew is already installed)

sarakonradi replied:
Done.

terhorstd commented on doc/documentation_workflow/user_documentation_workflow.rst:
```diff
@@ -0,0 +1,130 @@
+
+.. code-block:: bash
+
+   make html
```
The environment does not contain the required breathe extension
```
Extension error:
Could not import extension breathe (exception: No module named 'breathe')
Makefile:20: recipe for target 'html' failed
```

sarakonradi replied:
See comment above

sarakonradi commented on doc/documentation_workflow/user_documentation_workflow.rst:
```diff
@@ -0,0 +1,130 @@
+
+.. code-block:: bash
+
+   conda env create -f nest_doc_conda_env.yml
```
@terhorstd The old environment had been updated to ``nest_doc_conda_env.yml`` after the merge of #1411: https://github.com/nest/nest-simulator/blob/master/doc/nest_doc_conda_env.yml

sarakonradi replied:
@terhorstd I updated this branch. It now includes the merge of #1411, which updates the conda environment and renames it to ``nest_doc_conda_env.yml``. You should now be able to see the file!

terhorstd replied:
Thanks! Creating the envrironment works now.
I would still suggest to use long options, at least for the filename.

sarakonradi commented on doc/documentation_workflow/user_documentation_workflow.rst:
```diff
@@ -0,0 +1,130 @@
+
+   conda env create -f nest_doc_conda_env.yml
+   conda update -n base -c defaults conda
+   conda activate nest-doc
```
See comment above

sarakonradi replied:
``nest-doc`` should appear now (see comment above).

terhorstd replied:
Thanks.

sarakonradi commented on doc/contents.rst:
```diff
@@ -28,6 +28,7 @@ Table of Contents
 
    community
    contribute/index
+   documentation_workflow/user_documentation_workflow
```
Added both to ``documentation_workflow/index.rst`` so that the user can choose.

sarakonradi commented on doc/documentation_workflow/user_documentation_workflow.rst:
```diff
@@ -0,0 +1,130 @@
+.. code-block:: bash
+
+   conda env create -f nest_doc_conda_env.yml
+   conda update -n base -c defaults conda
```
Rearranged the order.

terhorstd replied:
ok, we can keep in the good practice.

sarakonradi commented:
> Everything works as intended, and I also think that it is now clear how the process works.

@jasperalbers Great. I added the macOS instructions for ``brew``.

> I bumped into some difficulties here (see comments).

@terhorstd The branch is now up-to-date, and you should be able to see ``nest_doc_conda_env.yml`` and ``nest-doc``. It should also solve the ``breathe`` issue. Let me know if it works now!

I added a "Documentation workflows" index to the navigation, which allows you to choose between user and developer documentation.

terhorstd commented on doc/documentation_workflow/user_documentation_workflow.rst:
```diff
@@ -0,0 +1,128 @@
+Creating pull request
++++++++++++++++++++++
+
+When you feel your documentation work is finished, you can create a pull request to the ``master`` branch of the NEST Source Code Repository. Your pull request will be reviewed by our NEST Documentation Team!
```
I think we have a description of how to do a PR against NEST and the development workflow. I would have expected a link here to see where to go...

sarakonradi replied:
Yes, there is a description in the NEST developer space. Added the link.

terhorstd commented on doc/documentation_workflow/developer_documentation_workflow.rst:
```diff
@@ -0,0 +1,79 @@
+.. code-block::
+
+   cd doc/doxygen/html
+   browser filename.html
```
this should probably also be index.html
```suggestion
   browser index.html
```

sarakonradi replied:
Done.

terhorstd commented on doc/documentation_workflow/developer_documentation_workflow.rst:
```diff
@@ -0,0 +1,79 @@
+
+    The ``helpdesk()`` command does not work if you have compiled
+    NEST with MPI support, but you have to enter the address of the helpdesk
+    (``file:///</path/to/nest_install_dir>/share/doc/nest/index.html``) manually into the browser.
+    Replace ``/install/path`` with the path under which NEST is installed.
```
This should be consistent
```suggestion
    (``file:///</install/path>/share/doc/nest/index.html``) manually into the browser.
    Replace ``</install/path>`` with the path under which NEST is installed.
```

sarakonradi replied:
Done.

sarakonradi commented:
@terhorstd @heplesser I updated the files based on your suggestions. Could you check if I addressed everything?

sarakonradi commented:
@heplesser No problem! I removed the old workflow graphic from ``_static/img``, in case that would lead to a Sphinx error.

terhorstd commented on doc/documentation_workflow/user_documentation_workflow.rst:
```diff
@@ -0,0 +1,130 @@
+.. code-block:: bash
+
+   conda update -n base -c defaults conda
+   conda env create -f nest_doc_conda_env.yml
```
```suggestion
   conda env create --file nest_doc_conda_env.yml
```
